File: src/client/python/agentguard/harness/runtime.py
"""HarnessRuntime: orchestrates the full client-side execution flow."""
from __future__ import annotations

from typing import Any, Callable

from agentguard.audit.recorder import AuditRecorder
from agentguard.harness.event_bus import EventBus
from agentguard.harness.lifecycle import Lifecycle
from agentguard.harness.session import Session
from agentguard.interceptors import (
    InputInterceptor,
    LLMInterceptor,
    MemoryInterceptor,
    OutputInterceptor,
    ThoughtInterceptor,
    ToolInterceptor,
    ToolResultInterceptor,
)
from agentguard.parser.output_router import OutputKind, route_output
from agentguard.parser.repair import repair_tool_call
from agentguard.sandbox.executor import SandboxExecutor
from agentguard.schemas import events as ev
from agentguard.schemas.context import RuntimeContext
from agentguard.schemas.decisions import DecisionType, GuardDecision
from agentguard.schemas.events import EventType, RuntimeEvent
from agentguard.tools.degrade import ToolDegradeManager
from agentguard.tools.metadata import ToolMetadata
from agentguard.tools.registry import ToolRegistry
from agentguard.u_guard.enforcer import EnforcementResult, UGuardEnforcer

_INTERCEPTORS = {
    EventType.USER_INPUT: InputInterceptor(),
    EventType.LLM_INPUT: LLMInterceptor(),
    EventType.LLM_OUTPUT: LLMInterceptor(),
    EventType.LLM_THOUGHT: ThoughtInterceptor(),
    EventType.TOOL_INVOKE: ToolInterceptor(),
    EventType.TOOL_RESULT: ToolResultInterceptor(),
    EventType.FINAL_RESPONSE: OutputInterceptor(),
    EventType.MEMORY_READ: MemoryInterceptor(),
    EventType.MEMORY_WRITE: MemoryInterceptor(),
}

_HOOK_BY_TYPE = {
    EventType.LLM_INPUT: "on_llm_input",
    EventType.LLM_OUTPUT: "on_llm_output",
    EventType.LLM_THOUGHT: "on_llm_thought",
    EventType.TOOL_INVOKE: "on_tool_invoke",
    EventType.TOOL_RESULT: "on_tool_result",
}


class HarnessRuntime:
    def __init__(
        self,
        *,
        context: RuntimeContext,
        enforcer: UGuardEnforcer,
        sandbox: SandboxExecutor,
        audit: AuditRecorder,
        registry: ToolRegistry | None = None,
        degrade_manager: ToolDegradeManager | None = None,
        lifecycle: Lifecycle | None = None,
        event_bus: EventBus | None = None,
        max_steps: int = 12,
        max_tool_calls: int = 24,
        window_size: int = 8,
    ) -> None:
        self.context = context
        self.enforcer = enforcer
        self.sandbox = sandbox
        self.audit = audit
        self.registry = registry or ToolRegistry()
        self.degrade = degrade_manager or ToolDegradeManager()
        self.lifecycle = lifecycle or Lifecycle()
        self.bus = event_bus or EventBus()
        self.max_steps = max_steps
        self.max_tool_calls = max_tool_calls
        self.window_size = window_size
        self.session = Session(context=context)
        # Share the session trace with the audit recorder for one history.
        self.audit.trace = self.session.trace
        self.enforcer.trace_window_provider = lambda: self.session.trace.window(window_size)

    # ---- event plumbing ------------------------------------------------
    def _intercept(self, event: RuntimeEvent, phase: str) -> RuntimeEvent:
        interceptor = _INTERCEPTORS.get(event.event_type)
        if interceptor is None:
            return event
        return interceptor.before(event, self.context) if phase == "before" else interceptor.after(
            event, self.context
        )

    def guard(
        self, event: RuntimeEvent, *, force_remote: bool = False, phase: str = "before"
    ) -> EnforcementResult:
        """Run interceptors, plugin hooks, enforcement and audit for an event."""
        event = self._intercept(event, phase)
        self.lifecycle.dispatch("on_event", event, self.context)
        hook = _HOOK_BY_TYPE.get(event.event_type)
        if hook:
            self.lifecycle.dispatch(hook, event, self.context)

        ext = self._collect_extensions(event)
        result = self.enforcer.enforce(
            event, self.context, plugin_extensions=ext, force_remote=force_remote
        )
        if result.route == "remote":
            self.lifecycle.dispatch(
                "on_after_remote_decision", result.decision, self.context
            )
        plugin_results = result.decision.metadata.get("plugin_results") or {}
        self.audit.record(event, result.decision, plugin_results)
        self.bus.publish(event)
        return result

    def _collect_extensions(self, event: RuntimeEvent) -> dict[str, Any]:
        request = {
            "plugin_extensions": {},
            "trajectory_window": [e.to_dict() for e in self.session.trace.window(self.window_size)],
            "event": event.to_dict(),
        }
        out = self.lifecycle.dispatch("on_before_remote_decision", request, self.context)
        return (out or {}).get("plugin_extensions", {})

    # ---- tool flow -----------------------------------------------------
    def invoke_tool(
        self,
        *,
        tool_name: str,
        arguments: dict[str, Any],
        fn: Callable[..., Any],
        metadata: ToolMetadata | None = None,
    ) -> Any:
        meta = metadata or self.registry.metadata(tool_name) or ToolMetadata(name=tool_name)
        if self.session.tool_call_count >= self.max_tool_calls:
            return self._safe_error("tool call budget exceeded", tool_name)
        self.session.inc_tool_call()

        invoke_event = ev.tool_invoke(
            self.context, tool_name, arguments, capabilities=list(meta.capabilities)
        )
        result = self.guard(invoke_event)
        decision = result.decision

        if decision.decision_type == DecisionType.DENY:
            return self._safe_error(decision.reason, tool_name, decision)
        if decision.requires_user or decision.requires_remote:
            return self._pending(decision.reason, tool_name, decision)
        if decision.decision_type == DecisionType.DEGRADE:
            return self._run_degraded(tool_name, arguments, decision)

        return self._execute(tool_name, arguments, fn, list(meta.capabilities), decision)

    def _execute(
        self,
        tool_name: str,
        arguments: dict[str, Any],
        fn: Callable[..., Any],
        capabilities: list[str],
        invoke_decision: GuardDecision,
    ) -> Any:
        sb = self.sandbox.run(fn, arguments, capabilities=capabilities, tool_name=tool_name)
        if not sb.success:
            err_event = ev.tool_result(self.context, tool_name, None, error=sb.error)
            self.guard(err_event, phase="after")
            return self._safe_error(sb.error or "tool failed", tool_name)

        result_event = ev.tool_result(self.context, tool_name, sb.value)
        res = self.guard(result_event, phase="after")
        rd = res.decision
        if rd.decision_type == DecisionType.DENY:
            return self._safe_error(rd.reason, tool_name, rd)
        if rd.decision_type == DecisionType.SANITIZE:
            return {"agentguard": "sanitized", "reason": rd.reason, "tool": tool_name}
        if rd.requires_user or rd.requires_remote:
            return self._pending(rd.reason, tool_name, rd)
        return sb.value

    def _run_degraded(
        self, tool_name: str, arguments: dict[str, Any], decision: GuardDecision
    ) -> Any:
        plan = self.degrade.plan(tool_name, arguments, decision.reason)
        if not plan.degraded or not plan.target_tool:
            return self._safe_error(plan.safe_error or "degradation failed", tool_name, decision)
        target = self.registry.get(plan.target_tool)
        if target is None:
            return {
                "agentguard": "degraded",
                "tool": tool_name,
                "degraded_to": plan.target_tool,
                "explanation": plan.explanation,
            }
        sb = self.sandbox.run(
            target.fn, plan.arguments, capabilities=list(target.metadata.capabilities),
            tool_name=plan.target_tool,
        )
        return sb.value if sb.success else self._safe_error(sb.error or "degraded tool failed", tool_name)

    # ---- llm output flow ----------------------------------------------
    def process_output(self, output: Any) -> dict[str, Any]:
        """Classify and guard a single LLM output. Returns a structured action."""
        routed = route_output(output)

        if routed.kind == OutputKind.THOUGHT_TRACE:
            event = ev.llm_thought(self.context, routed.thought or "")
            event.risk_signals.extend(routed.risk_signals)
            decision = self.guard(event).decision
            if decision.decision_type in (DecisionType.DROP_THOUGHT, DecisionType.DENY):
                return {"kind": "thought_dropped", "reason": decision.reason}
            return {"kind": "thought", "thought": routed.thought}

        if routed.kind == OutputKind.TOOL_CALL_CANDIDATE:
            return {"kind": "tool_calls", "tool_calls": routed.tool_calls}

        if routed.kind == OutputKind.MALFORMED_TOOL_CALL:
            return {"kind": "malformed", "errors": routed.errors}

        # final_response or unsafe_output
        event = ev.final_response(self.context, routed.text or "")
        event.risk_signals.extend(routed.risk_signals)
        decision = self.guard(event).decision
        if decision.decision_type == DecisionType.DENY:
            return {"kind": "final", "text": f"[AgentGuard blocked: {decision.reason}]", "blocked": True}
        if decision.decision_type == DecisionType.SANITIZE:
            return {"kind": "final", "text": "[AgentGuard sanitized output]", "sanitized": True}
        return {"kind": "final", "text": routed.text}

    def run_agent(self, adapter: Any, agent: Any, input_data: Any) -> dict[str, Any]:
        """Drive a guarded ReAct loop using an agent adapter."""
        ui = ev.user_input(self.context, str(input_data))
        self.guard(ui)
        messages: list[dict[str, Any]] = [{"role": "user", "content": str(input_data)}]
        last_final: str | None = None

        for _ in range(self.max_steps):
            self.session.inc_step()
            self.guard(ev.llm_input(self.context, list(messages)))
            output = adapter.generate(agent, messages, self.context)
            self.guard(ev.llm_output(self.context, output))
            action = self.process_output(output)

            if action["kind"] == "tool_calls":
                for tc in action["tool_calls"]:
                    obs = self._invoke_parsed(tc)
                    messages.append({"role": "tool", "name": tc.tool_name, "content": str(obs)})
                continue
            if action["kind"] in ("thought", "thought_dropped"):
                messages.append({"role": "assistant", "content": str(action.get("thought", ""))})
                continue
            if action["kind"] == "malformed":
                messages.append({"role": "user", "content": "Your tool call was malformed; retry."})
                continue
            last_final = action.get("text")
            break

        return {"final": last_final, "steps": self.session.step_count, "trace": self.session.trace}

    def _invoke_parsed(self, tool_call: Any) -> Any:
        reg = self.registry.get(tool_call.tool_name)
        if reg is None:
            repaired = repair_tool_call(tool_call, known_tools=self.registry.names())
            if not repaired.success or repaired.tool_call is None:
                return self._safe_error(f"unknown tool '{tool_call.tool_name}'", tool_call.tool_name)
            reg = self.registry.get(repaired.tool_call.tool_name)
            tool_call = repaired.tool_call
        if reg is None:
            return self._safe_error("tool not registered", tool_call.tool_name)
        return self.invoke_tool(
            tool_name=tool_call.tool_name,
            arguments=tool_call.arguments,
            fn=reg.fn,
            metadata=reg.metadata,
        )

    # ---- safe results --------------------------------------------------
    @staticmethod
    def _safe_error(reason: str, tool: str, decision: GuardDecision | None = None) -> dict[str, Any]:
        return {
            "agentguard": "blocked",
            "tool": tool,
            "reason": reason,
            "decision": decision.decision_type.value if decision else "deny",
        }

    @staticmethod
    def _pending(reason: str, tool: str, decision: GuardDecision) -> dict[str, Any]:
        return {
            "agentguard": "pending",
            "tool": tool,
            "reason": reason,
            "decision": decision.decision_type.value,
        }

File: src/client/js/agentguard/harness/runtime.js
"use strict";

const ev = require("../schemas/events");
const { DecisionType, GuardDecision } = require("../schemas/decisions");
const { Session } = require("./session");
const { EventBus } = require("./event_bus");
const { Lifecycle } = require("./lifecycle");
const { ToolRegistry } = require("../tools/registry");
const { ToolMetadata } = require("../tools/metadata");
const { ToolDegradeManager } = require("../tools/degrade");
const { LLMInterceptor, ToolInterceptor, ToolResultInterceptor } = require("../interceptors");

const INTERCEPTORS = {
  llm_input: new LLMInterceptor(),
  llm_output: new LLMInterceptor(),
  tool_invoke: new ToolInterceptor(),
  tool_result: new ToolResultInterceptor(),
};

const HOOK_BY_TYPE = {
  llm_input: "on_llm_input",
  llm_output: "on_llm_output",
  tool_invoke: "on_tool_invoke",
  tool_result: "on_tool_result",
};

class HarnessRuntime {
  constructor({
    context,
    enforcer,
    sandbox,
    audit,
    registry = null,
    degrade_manager = null,
    lifecycle = null,
    event_bus = null,
    max_steps = 12,
    max_tool_calls = 24,
    window_size = 8,
  }) {
    this.context = context;
    this.enforcer = enforcer;
    this.sandbox = sandbox;
    this.audit = audit;
    this.registry = registry || new ToolRegistry();
    this.degrade = degrade_manager || new ToolDegradeManager();
    this.lifecycle = lifecycle || new Lifecycle();
    this.bus = event_bus || new EventBus();
    this.max_steps = max_steps;
    this.max_tool_calls = max_tool_calls;
    this.window_size = window_size;
    this.session = new Session({ context });
    this.audit.trace = this.session.trace;
    this.enforcer.trace_window_provider = () => this.session.trace.window(window_size);
  }

  intercept(event, phase) {
    const interceptor = INTERCEPTORS[event.event_type];
    if (!interceptor) {
      return event;
    }
    return phase === "before" ? interceptor.before(event, this.context) : interceptor.after(event, this.context);
  }

  async guard(event, { force_remote = false, phase = "before" } = {}) {
    const nextEvent = this.intercept(event, phase);
    this.lifecycle.dispatch("on_event", nextEvent, this.context);
    const hook = HOOK_BY_TYPE[nextEvent.event_type];
    if (hook) {
      this.lifecycle.dispatch(hook, nextEvent, this.context);
    }
    const result = await this.enforcer.enforce(nextEvent, this.context, { force_remote });
    this.audit.record(nextEvent, result.decision);
    this.bus.publish(nextEvent);
    return result;
  }

  async invoke_tool({ tool_name, arguments: arguments_, fn, metadata = null }) {
    try {
      return await this.invokeToolInner({ tool_name, arguments: arguments_, fn, metadata });
    } catch (error) {
      await this.sync_local_cache_now({ reason: "client_error" });
      throw error;
    } finally {
      this.sync_local_cache_async({ reason: "round_complete" });
    }
  }

  async invokeToolInner({ tool_name, arguments: arguments_, fn, metadata = null }) {
    const meta = metadata || this.registry.metadata(tool_name) || new ToolMetadata({ name: tool_name });
    if (this.session.tool_call_count >= this.max_tool_calls) {
      return this.safeError("tool call budget exceeded", tool_name);
    }
    this.session.inc_tool_call();
    const invokeEvent = ev.tool_invoke(this.context, tool_name, arguments_, {
      capabilities: [...(meta.capabilities || [])],
    });
    const result = await this.guard(invokeEvent);
    const decision = result.decision;
    if (decision.decision_type === DecisionType.DENY) {
      return this.safeError(decision.reason, tool_name, decision);
    }
    if (decision.requires_user || decision.requires_remote) {
      return this.pending(decision.reason, tool_name, decision);
    }
    if (decision.decision_type === DecisionType.DEGRADE) {
      return this.runDegraded(tool_name, arguments_, decision);
    }
    return this.execute(tool_name, arguments_, fn, [...(meta.capabilities || [])]);
  }

  sync_local_cache_async({ reason = "round_complete" } = {}) {
    const remote = this.enforcer.remote;
    const buffer = this.enforcer.sync_buffer;
    if (!remote || !remote.enabled || !buffer || !buffer.has_entries()) {
      return false;
    }
    const entries = buffer.snapshot();
    if (!entries.length) {
      return false;
    }
    const trace = buffer.build_trace_upload({
      context: this.context,
      entries,
      reason,
    });
    remote.upload_trace_async(trace, {
      on_success: () => buffer.remove_entries(entries),
    });
    return true;
  }

  async sync_local_cache_now({ reason = "client_error" } = {}) {
    const remote = this.enforcer.remote;
    const buffer = this.enforcer.sync_buffer;
    if (!remote || !remote.enabled || !buffer || !buffer.has_entries()) {
      return false;
    }
    const entries = buffer.pop_all();
    if (!entries.length) {
      return false;
    }
    const trace = buffer.build_trace_upload({
      context: this.context,
      entries,
      reason,
    });
    try {
      await remote.upload_trace(trace);
      return true;
    } catch (_) {
      buffer.restore_front(entries);
      return false;
    }
  }

  async execute(toolName, arguments_, fn, capabilities) {
    const sandboxResult = this.sandbox.run(fn, arguments_, {
      capabilities,
      tool_name: toolName,
    });
    const resolved = sandboxResult && typeof sandboxResult.then === "function" ? await sandboxResult : sandboxResult;
    if (!resolved.success) {
      const errorEvent = ev.tool_result(this.context, toolName, null, { error: resolved.error });
      await this.guard(errorEvent, { phase: "after" });
      return this.safeError(resolved.error || "tool failed", toolName);
    }
    const resultEvent = ev.tool_result(this.context, toolName, resolved.value);
    const guardResult = await this.guard(resultEvent, { phase: "after" });
    const decision = guardResult.decision;
    if (decision.decision_type === DecisionType.DENY) {
      return this.safeError(decision.reason, toolName, decision);
    }
    if (decision.decision_type === DecisionType.SANITIZE) {
      return { agentguard: "sanitized", reason: decision.reason, tool: toolName };
    }
    if (decision.requires_user || decision.requires_remote) {
      return this.pending(decision.reason, toolName, decision);
    }
    return resolved.value;
  }

  runDegraded(toolName, arguments_, decision) {
    const plan = this.degrade.plan(toolName, arguments_, decision.reason);
    if (!plan.degraded || !plan.target_tool) {
      return this.safeError(plan.safe_error || "degradation failed", toolName, decision);
    }
    const target = this.registry.get(plan.target_tool);
    if (!target) {
      return {
        agentguard: "degraded",
        tool: toolName,
        degraded_to: plan.target_tool,
        explanation: plan.explanation,
      };
    }
    const sandboxResult = this.sandbox.run(target.fn, plan.arguments, {
      capabilities: [...(target.metadata.capabilities || [])],
      tool_name: plan.target_tool,
    });
    return sandboxResult.success ? sandboxResult.value : this.safeError(sandboxResult.error || "degraded tool failed", toolName);
  }

  safeError(reason, tool, decision = null) {
    return {
      agentguard: "blocked",
      tool,
      reason,
      decision: decision ? decision.decision_type : GuardDecision.deny(reason).decision_type,
    };
  }

  pending(reason, tool, decision) {
    return {
      agentguard: "pending",
      tool,
      reason,
      decision: decision.decision_type,
    };
  }
}

module.exports = {
  HarnessRuntime,
};
